### PR TITLE
Vlan pool support for vlan by id

### PR DESCRIPTION
**Pull request by fdurand**
# Description

Adds a way to define a pool of vlan to use in a role.
# Impacts

No
# Delete branch after merge

YES
# NEWS file entries
## New Features
- Added support for vlan pool in role by vlan id

### CODE REVIEW

File: lib/pf/vlan/pool.pm [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+}
+
+
+=head2 getRoundRobbin
+
+Return the vlan id based on round robbin
+
+=cut
+
+sub getRoundRobbin {
```
Comment (julsemaan) [inline lib/pf/vlan/pool.pm]: *getRoundRobin

It's spelled with only 1 b

File: lib/pf/vlan/pool.pm [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+         $rangesep$
+         )/x;
+    return 1 if ($range =~ m/$validation/g);
+    return 0;
+}
+
+
+=head2 getRoundRobbin
+
+Return the vlan id based on round robbin
```
Comment (julsemaan) [inline lib/pf/vlan/pool.pm]: *robin

File: lib/pf/vlan/pool.pm [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+         ^$rangesep|
+         $sectsep$|
+         $rangesep$
+         )/x;
+    return 1 if ($range =~ m/$validation/g);
+    return 0;
+}
+
+
+=head2 getRoundRobbin
```
Comment (julsemaan) [inline lib/pf/vlan/pool.pm]: getRoundRobin

File: lib/pf/vlan/pool.pm [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+}
+
+sub getVlanPool {
+    my ($self, $vlan, $switch, $ifIndex, $mac, $node_info, $connection_type, $user_name, $ssid, $radius_request) = @_;
+    my $logger =  pf::log::get_logger();
+
+    return $vlan if $self->rangeValidator($vlan);
+    my $range = Number::Range->new($vlan);
+
+    $vlan = $self->getRoundRobbin($mac, $node_info, $range);
```
Comment (julsemaan) [inline lib/pf/vlan/pool.pm]: getRoundRobin

File: lib/pf/vlan.pm [hunk 5]
```diff
@@ -496,6 +497,8 @@ sub getNormalVlan {
     }
     $vlan = $switch->getVlanByName($role);
     $pf::StatsD::statsd->end(called() . ".timing" , $start, 0.05 );
+    my $vlanpool = new pf::vlan::pool;
+    $vlan = $vlanpool->getVlanPool($vlan, $switch, $ifIndex, $mac, $node_info, $connection_type, $user_name, $ssid, $radius_request);
```
Comment (julsemaan) [inline lib/pf/vlan.pm]: These two lines should go above the timing as now there's not timed at all

File: lib/pf/vlan/pool.pm [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+sub new {
+    my $logger =  pf::log::get_logger();
+    $logger->debug("instantiating new pf::vlan::pool object");
+    my ( $class, %argv ) = @_;
+    my $this = bless {}, $class;
+    return $this;
+}
+
+sub getVlanPool {
```
Comment (louismunro) [inline lib/pf/vlan/pool.pm]: getVlanFromPool? 
Otherwise it sounds like it is returning the whole pool.

File: lib/pf/vlan/pool.pm [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+
+Validate the range definition
+Should be something like that 20..23 or 20..23,27..30
+
+=cut
+
+sub rangeValidator {
+    my ($self, $range) =@_;
+    my $rangesep = qr/(?:\.\.)/;
+    my $sectsep  = qr/(?:\s|,)/;
```
Comment (louismunro) [inline lib/pf/vlan/pool.pm:67]: I am not sure we should accept spaces as section separators.
It may be much easier in the end to just ignore spaces.
That could prevent a whole class of hard to debug errors (typos etc.)

File: lib/pf/vlan/pool.pm [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+    return 0;
+}
+
+
+=head2 getRoundRobin
+
+Return the vlan id based on round robin
+
+=cut
+
```
Comment (louismunro) [inline lib/pf/vlan/pool.pm:90]: This code needs some explanations. It is a bit hard to follow.

File: lib/pf/vlan/pool.pm [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+    my ($self, $mac, $role, $range) = @_;
+    my $logger =  pf::log::get_logger();
+
+    my $vlan_count = $range->size;
+    my $node_info_complete = node_view($mac);
+    if ( defined($node_info_complete->{'last_vlan'}) && $range->inrange($node_info_complete->{'last_vlan'}) ) {
+        $logger->debug("NODE LAST VLAN ".$node_info_complete->{'last_vlan'});
+        return ($node_info_complete->{'last_vlan'});
+    }
+    my $last_reg_mac = node_last_reg_non_inline_on_category($mac, $role);
```
Comment (louismunro) [inline lib/pf/vlan/pool.pm:101]: Performance worries me here.
Joins tend to be bad when it comes time to scale.

Comment (extrafu) [inline lib/pf/vlan/pool.pm:101]: Why not simply use a robust hash function? The bucket size could be configurable.

File: lib/pf/vlan/pool.pm [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+    return $this;
+}
+
+sub getVlanPool {
+    my ($self, $vlan, $switch, $ifIndex, $mac, $node_info, $connection_type, $user_name, $ssid, $radius_request, $role) = @_;
+    my $logger =  pf::log::get_logger();
+
+    return $vlan if $self->rangeValidator($vlan);
+    my $range = Number::Range->new($vlan);
+
```
Comment (louismunro) [inline lib/pf/vlan/pool.pm:52]: Not sure I understand this part.
If the rangeValidator returns 1 we will never hit the getRoundRobin() ?

File: lib/pf/vlan.pm [hunk 11]
```diff
@@ -495,6 +496,8 @@ sub getNormalVlan {
         $logger->info("[$mac] Username was NOT defined or unable to match a role - returning node based role '$role'");
     }
     $vlan = $switch->getVlanByName($role);
+    my $vlanpool = new pf::vlan::pool;
+    $vlan = $vlanpool->getVlanPool($vlan, $switch, $ifIndex, $mac, $node_info, $connection_type, $user_name, $ssid, $radius_request, $role);
```
Comment (jrouzierinverse) [inline lib/pf/vlan.pm]: Change the call to submit a single hash ref instead of the long list of parameters.
$vlanpool->getVlanPool({vlan => $vlan, switch =>  $switch, ifIndex => $ifIndex, mac => $mac,  node_info => $node_info, connection_type => $connection_type, user_name => $user_name, ssid => $ssid, radius_request => $radius_request, role => $role});

Comment (dwlfrth) [inline lib/pf/vlan.pm]: I'd build the hash separately to keep it clean and then pass the reference in the call :) :nail_care:

Comment (jrouzierinverse) [inline lib/pf/vlan.pm]: my $pool_args = {vlan => $vlan, switch => $switch, ifIndex => $ifIndex, mac => $mac, node_info => $node_info, connection_type => $connection_type, user_name => $user_name, ssid => $ssid, radius_request => $radius_request, role => $role};
$vlan = $vlanpool->getVlanPool($pool_args);

It makes it easier to add new parameters later

Comment (dwlfrth) [inline lib/pf/vlan.pm]: Exaaaaccccttlllyyyy !!

``` perl
my $pool_args = {
    vlan => $vlan,
    switch => $switch,
    ifIndex => $ifIndex,
    mac => $mac,
    node_info => $node_info,
    connection_type => $connection_type,
    user_name => $user_name,
    ssid => $ssid,
    radius_request => $radius_request,
    role => $role,
};
$vlan = $vlanpool->getVlanPool($pool_args);
```

"I'm sexy and I know it"

File: lib/pf/vlan/pool.pm (hunk 10 above)
Comment (fdurand) [inline lib/pf/vlan/pool.pm:52]: if it return 1 it mean that it's not a range

File: lib/pf/vlan/pool.pm (hunk 9 above)
Comment (fdurand) [inline lib/pf/vlan/pool.pm:101]: The last reg non inline on category value must be shared between all the radius process.
The better way is to take it from the database

File: lib/pf/vlan/pool.pm (hunk 7 above)
Comment (fdurand) [inline lib/pf/vlan/pool.pm:67]: This code is based on the Number::Range module http://cpansearch.perl.org/src/LARRYSH/Number-Range-0.12/lib/Number/Range.pm it's use to detect that it's a range of number and not only a number or string

File: lib/pf/vlan.pm (hunk 11 above)
Comment (fdurand) [inline lib/pf/vlan.pm]: i don't like hash

Comment (dwlfrth) [inline lib/pf/vlan.pm]: And I don't like having to modify all the code to add a parameter :P ahah

File: lib/pf/vlan/pool.pm (hunk 9 above)
Comment (fdurand) [inline lib/pf/vlan/pool.pm:101]: Btw it will only happen one time on registration. If the device is reg then we use the last_vlan instead.

**Comment by fdurand** [comment]
I want to rework it because i refactor the vlan.pm in another branch.

**Comment by louismunro** [comment]
Ok, we really should sit down together sometime this week and see that we can merge both of our implementations.

**Comment by louismunro** [comment]
Closing for now.
We will reopen.
